### PR TITLE
PERF: Improve Styler `to_excel` Performance

### LINKED ISSUE #47352: PERF: Excel Styler treatment of CSS side expansion is slow
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this issue exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [x] I have confirmed this issue exists on the main branch of pandas.

### Reproducible Example

The styler approach to resolving CSS string expansion for conversion to Excel is slow. In particular:

- `atomize` contains an inefficient string addition and replace.
- No caching is conducted by `CSSResolver` or `CSSToExcelConverter`

```
from pathlib import Path

import numpy as np
import pandas as pd

df = pd.DataFrame(np.random.randint(0, 100, (10, 1000)))
style = df.style
style.applymap(lambda x: "border-color: black; border-style: solid; border-width: thin")
style.to_excel(Path(__file__).parent / "test.xlsx")
```

On my machine, this example took 12.2s to run. Of that time, 6.57s was cumulatively spent in `css.py:CSSResolver.__call__`, including 2.37s in `parse` and 1.5s in `atomize`.

### Installed Versions

<details>

Replace this line with the output of pd.show_versions()

</details>

### Prior Performance

_No response_

### PR DESCRIPTION
**Pull request by tehunter**
- [x] closes #47352
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by tehunter** [comment]
ASV Results:

```
..       before           after         ratio
     [6f0be79b]       [db365209]
                      <styler-performance>
-       915±200ms         594±30ms     0.65  io.excel.WriteExcel.time_write_excel('openpyxl')
-       658±200ms         323±20ms     0.49  io.excel.WriteExcel.time_write_excel('xlsxwriter')
-       3.08±0.1s         934±40ms     0.30  io.excel.WriteExcelStyled.time_write_excel_style('xlsxwriter')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

**Comment by tehunter** [comment]
@github-actions pre-commit

**Comment by tehunter** [comment]
Overview of changes:

- Moved `CSSResolver.expand_*` methods (which are accessed in `.atomize`) to a dictionary called `CSS_EXPANSIONS` which is keyed directly by the CSS property, instead of having to do a string replace and concatenation.
- `CSSResolver.__call__` now accepts an iterable of declaration tuples `(property, value)` in addition to the previous string type.
    - When an iterable is passed, `.parse` does not need to be called
- `CSSExcelCell.__init__` no longer converts the css_styles (i.e. `Styler.ctx`) from a list of tuples to a CSS string. The list of tuples was already obtained from the initial CSS string(s) in `Styler._update_ctx`, so we were previously parsing, reforming into string, and parsing again. Instead, the list is converted to a frozenset via a dictionary (to keep only the final instance of each property declaration)
- `CSSToExcelConverter.__call__` is now cached using `lru_cache`. The frozenset was necessary to make all the arguments hashable for the caching mechanism.

File: pandas/io/formats/excel.py [hunk 1]
```diff
@@ -85,10 +89,10 @@ def __init__(
         **kwargs,
     ) -> None:
         if css_styles and css_converter:
-            css = ";".join(
-                [a + ":" + str(v) for (a, v) in css_styles[css_row, css_col]]
-            )
-            style = css_converter(css)
+            # Use dict to get only one declaration per property, then convert
+            # to frozenset for caching
+            unique_declarations = frozenset(dict(css_styles[css_row, css_col]).items())
```
Comment (mroeschke) [inline pandas/io/formats/excel.py:94]: Maybe safer to use a tuple instead of a frozenset to guarantee the ordering for a consistent cache key?

Comment (tehunter) [reply pandas/io/formats/excel.py]: I think we actually want it to be order-invariant here. Here's my reasoning:

1. The `dict(...)` ensures that each unique property is only present once; specifically, the final declaration in the CSS string for each property is the one that is stored in the dict.
2. The `.items()` call converts that dictionary into `(property, value)` pairs. By (1), this collection of pairs is guaranteed to have each property show up only once.
3. The frozenset has the same hash iff every tuple in the collection of pairs is the same, with no concept of order. Since the set can't contain duplicate properties, then two sets with the same hash must represent the same declarations. For example, if in Cell A1 I applied `font-color: red; font-weight: bold` and Cell B1 I applied `font-weight: bold; font-color: red`, they can still be safely cached as one.

I added another set of commits that added tests to ensure it's working as intended. Please take a look. If my logic isn't sound or I'm overlooking any edge cases, please let me know!

File: asv_bench/benchmarks/io/excel.py [hunk 2]
```diff
@@ -47,6 +47,25 @@ def time_write_excel(self, engine):
...
 
 
+class WriteExcelStyled:
+    params = ["openpyxl", "xlsxwriter"]
+    param_names = ["engine"]
+
+    def setup(self, engine):
+        self.df = _generate_dataframe()
+
+    def time_write_excel_style(self, engine):
```
[the remark this answers is not in the document]
Comment (tehunter) [reply asv_bench/benchmarks/io/excel.py:57]: I'm not sure I follow. The benchmark should include cache hits because the cached function is run for each cell, so the example should have 3 misses and `(N*C - 3)` hits. I added a test in [8600ba6](https://github.com/pandas-dev/pandas/pull/47371/commits/8600ba641956eab195cd305ffab7cb4ceb208b09) that ensures the cache is actually being hit.

**Comment by tehunter** [comment]
Benchmarks as of [8e56402](https://github.com/pandas-dev/pandas/pull/47371/commits/8e56402ed0639c231c98d06c25c62ee3da6f4524)
```
.       before           after         ratio
     [2b1184dd]       [8e56402e]
                      <styler-performance>
-        260±10ms          233±5ms     0.90  io.excel.WriteExcel.time_write_excel('xlsxwriter')
-        620±70ms         450±20ms     0.73  io.excel.WriteExcel.time_write_excel('openpyxl')
-      2.51±0.02s        1.42±0.3s     0.57  io.excel.WriteExcelStyled.time_write_excel_style('openpyxl')
-      1.69±0.07s         608±10ms     0.36  io.excel.WriteExcelStyled.time_write_excel_style('xlsxwriter')
- ```

**Review by mroeschke — APPROVED** [review]
